typing/ctype.ml

```diff
@@ -1210,20 +1215,21 @@ let instance_list schl =
   For_copy.with_scope (fun copy_scope ->
     List.map (fun t -> copy copy_scope t) schl)
 
-let reified_var_counter = ref Vars.empty
-let reset_reified_var_counter () =
-  reified_var_counter := Vars.empty
-
-(* names given to new type constructors.
-   Used for existential types and
-   local constraints *)
-let get_new_abstract_name s =
-  let index =
-    try Vars.find s !reified_var_counter + 1
-    with Not_found -> 0 in
-  reified_var_counter := Vars.add s index !reified_var_counter;
-  if index = 0 && s <> "" && s.[String.length s - 1] <> '$' then s else
-  Printf.sprintf "%s%d" s index
+(* Create unique names to new type constructors.
+   Used for existential types and local constraints. *)
+let get_new_abstract_name env s =
+  (* unique names are needed only for error messages *)
+  if in_counterexample () then s else
```

Comment:
This assumes that `umode` is correctly set when this function is called (that is, if I trace things correctly, inside `Typecore.solve_Ppat_construct`), and not a leftover global value from a previous `subst` call or something. How do you know that `umode` is correctly set at this point?

Reply:
Because `umode` is only set using `Misc.protect_refs`, which guarantees that it will be returned to its original value.
Actually, looking carefully there is one occurence of `umode` being set directly in `Ctype.subst`, and it might not be reverted if an exception other than `Unify` occurs. In that case the leftover value would be `Subst`, which is fine for us, but should still probably be corrected.
As a side note, the worse that can happen if we guess wrong here it is that the error message would be slightly harder to read (it starts talking about location of internally generated variables).

Reply:
I complained about `umode` being global state that is mutated in unclear ways. @Octachron decided to volunteer to see if it can be turned into a parameter for the unification functions instead. (I think we should have a "context" record type with a `umode` field, so that we can add more things later easily).